File: shopify-app/address_privacy.py
"""Take street addresses out of any text the website shows (Naomi, 17 Sep 2026:
never show the address, suburb only; Instagram may keep its own).

    from address_privacy import redact
    redact("Meet Sierra. 📍 6 Dagmar Street, Holland Park West")
    # -> "Meet Sierra. 📍 Holland Park West"

The streets live in address-privacy.json (shared with address-privacy.mjs,
which must stay a line-for-line port of this file).
"""
import json, os, re

DATA = json.load(open(os.path.join(os.path.dirname(os.path.abspath(__file__)), 'address-privacy.json')))

def _either(word):
    return f"[{word[0].upper()}{word[0].lower()}]{re.escape(word[1:])}"

TYPES = '(?:' + '|'.join(x for t in sorted(DATA['types'], key=len, reverse=True) for x in (_either(t), re.escape(t.upper()))) + r')\b'
COLL = '(?:[Cc]ollection|COLLECTION)'
NUM = r'(\d+[A-Za-z]?)\s+'

def _cap(m, word, group=0):
    """Capitalise the replacement only where the original started a sentence."""
    before = m.string[:m.start()].rstrip()
    starts = not before or before[-1] in '.!?' or not before[-1].isalnum() and before[-1] not in ',;:'
    return word[0].upper() + word[1:] if m.group(group)[:1].isupper() and starts else word

def _street(text, st):
    name = '(?:' + r'\s+'.join(re.escape(w) for w in st['name'].split()) + '|' + r'\s+'.join(re.escape(w.upper()) for w in st['name'].split()) + ')'
    sub = re.escape(st['suburb'])
    houses = st['houses']
    sole = next(iter(houses.values())) if len(houses) == 1 else None
    # "the Newman Avenue Collection" -> "our Camp Hill collection"
    text = re.sub(rf"\b(?:[Tt]he|[Oo]ur)\s+{name}\s+{TYPES}\s+{COLL}", lambda m: _cap(m, f"our {st['suburb']} collection"), text)
    text = re.sub(rf"\b{name}\s+{TYPES}\s+{COLL}", lambda m: f"{st['suburb']} collection", text)
    # "on the corner of Buena Vista Avenue, Coorparoo" -> "on a corner in Coorparoo"
    text = re.sub(rf"\b([Oo]n|[Aa]t)\s+the\s+corner\s+of\s+(?:{NUM})?{name}\s+{TYPES}(?:,?\s+(?:in\s+)?{sub}\b)?",
                  lambda m: _cap(m, f"on a corner in {st['suburb']}", 1), text)
    # "on Newman Avenue in Camp Hill" / "at 29A Todd Street, Taringa" -> "in Camp Hill"
    text = re.sub(rf"\b([Oo]n|[Aa]long|[Aa]t)\s+(?:{NUM})?{name}\s+{TYPES}(?:,?\s+(?:in\s+)?{sub}\b)?",
                  lambda m: _cap(m, f"in {st['suburb']}", 1), text)
    # "26 Akala Street, Camp Hill" -> "MILOS, Camp Hill" (or just the suburb when the name is already there)
    def numbered(m):
        house = houses.get(m.group(1)) or houses.get(m.group(1).upper())
        before = text_ref[0][max(0, m.start() - 30):m.start()].lower()
        if house and house.lower() in before:
            house = None
        if m.group(2):
            return f"{house}, {st['suburb']}" if house else st['suburb']
        return house or st['suburb']
    text_ref = [text]
    text = re.sub(rf"{NUM}{name}\s+{TYPES}(,?\s+{sub}\b)?", numbered, text)
    # a bare "Arnold Street" -> "CASA PALMERA", or the suburb when the street had several houses
    text = re.sub(rf"\b{name}\s+{TYPES}", sole or st['suburb'], text)
    return text

UNKNOWN = re.compile(rf"\b\d+[A-Za-z]?(?:[-/]\d+)?\s+(?:[A-Z][a-z]+\s+){{1,3}}{TYPES},?\s*")

def redact(text, unknown=None):
    """Return text without street addresses. `unknown`, a list, collects any
    address that is not in address-privacy.json (those are removed too)."""
    if not text or not isinstance(text, str):
        return text
    for st in DATA['streets']:
        text = _street(text, st)
    def drop(m):
        if unknown is not None: unknown.append(m.group(0).strip())
        return ''
    return UNKNOWN.sub(drop, text)

def redact_json(value, unknown=None):
    """redact() every string inside a JSON value (lists, dicts)."""
    if isinstance(value, str): return redact(value, unknown)
    if isinstance(value, list): return [redact_json(v, unknown) for v in value]
    if isinstance(value, dict): return {k: redact_json(v, unknown) for k, v in value.items()}
    return value

File: shopify-app/address-privacy.mjs
/**
 * Take street addresses out of any text the website shows (Naomi, 17 Sep
 * 2026: never show the address, suburb only; Instagram may keep its own).
 * A line-for-line port of address_privacy.py; both read address-privacy.json.
 *
 *   import { redact } from './address-privacy.mjs';
 *   redact('Meet Sierra. 📍 6 Dagmar Street, Holland Park West');
 *   // -> 'Meet Sierra. 📍 Holland Park West'
 */
import fs from 'node:fs';
import path from 'node:path';
import { fileURLToPath } from 'node:url';

const DATA = JSON.parse(fs.readFileSync(path.join(path.dirname(fileURLToPath(import.meta.url)), 'address-privacy.json'), 'utf8'));
const esc = (s) => s.replace(/[.*+?^${}()|[\]\\]/g, '\\$&');
const either = (w) => `[${w[0].toUpperCase()}${w[0].toLowerCase()}]${esc(w.slice(1))}`;
const TYPES = '(?:' + [...DATA.types].sort((a, b) => b.length - a.length).flatMap((t) => [either(t), esc(t.toUpperCase())]).join('|') + ')\\b';
const COLL = '(?:[Cc]ollection|COLLECTION)';
const NUM = '(\\d+[A-Za-z]?)\\s+';
// Capitalise the replacement only where the original started a sentence.
const cap = (src, word, offset, whole) => {
  const before = whole.slice(0, offset).trimEnd();
  const last = before.slice(-1);
  const starts = !before || /[.!?]/.test(last) || (!/[\p{L}\p{N}]/u.test(last) && !/[,;:]/.test(last));
  return /^[A-Z]/.test(src) && starts ? word[0].toUpperCase() + word.slice(1) : word;
};

function street(text, st) {
  const words = st.name.split(/\s+/);
  const name = '(?:' + words.map(esc).join('\\s+') + '|' + words.map((w) => esc(w.toUpperCase())).join('\\s+') + ')';
  const sub = esc(st.suburb);
  const houses = st.houses;
  const vals = Object.values(houses);
  const sole = vals.length === 1 ? vals[0] : null;
  text = text.replace(new RegExp(`\\b(?:[Tt]he|[Oo]ur)\\s+${name}\\s+${TYPES}\\s+${COLL}`, 'g'), (m, offset, whole) => cap(m, `our ${st.suburb} collection`, offset, whole));
  text = text.replace(new RegExp(`\\b${name}\\s+${TYPES}\\s+${COLL}`, 'g'), () => `${st.suburb} collection`);
  text = text.replace(new RegExp(`\\b([Oo]n|[Aa]t)\\s+the\\s+corner\\s+of\\s+(?:${NUM})?${name}\\s+${TYPES}(?:,?\\s+(?:in\\s+)?${sub}\\b)?`, 'g'),
    (m, p, num, offset, whole) => cap(p, `on a corner in ${st.suburb}`, offset, whole));
  text = text.replace(new RegExp(`\\b([Oo]n|[Aa]long|[Aa]t)\\s+(?:${NUM})?${name}\\s+${TYPES}(?:,?\\s+(?:in\\s+)?${sub}\\b)?`, 'g'),
    (m, p, num, offset, whole) => cap(p, `in ${st.suburb}`, offset, whole));
  const src = text;
  text = text.replace(new RegExp(`${NUM}${name}\\s+${TYPES}(,?\\s+${sub}\\b)?`, 'g'), (m, num, withSuburb, offset) => {
    let house = houses[num] || houses[num.toUpperCase()] || null;
    if (house && src.slice(Math.max(0, offset - 30), offset).toLowerCase().includes(house.toLowerCase())) house = null;
    if (withSuburb) return house ? `${house}, ${st.suburb}` : st.suburb;
    return house || st.suburb;
  });
  text = text.replace(new RegExp(`\\b${name}\\s+${TYPES}`, 'g'), () => sole || st.suburb);
  return text;
}

const UNKNOWN = new RegExp(`\\b\\d+[A-Za-z]?(?:[-/]\\d+)?\\s+(?:[A-Z][a-z]+\\s+){1,3}${TYPES},?\\s*`, 'g');

/** Text without street addresses; `unknown` (an array) collects streets not in address-privacy.json (removed too). */
export function redact(text, unknown) {
  if (!text || typeof text !== 'string') return text;
  for (const st of DATA.streets) text = street(text, st);
  return text.replace(UNKNOWN, (m) => { if (unknown) unknown.push(m.trim()); return ''; });
}

/** redact() every string inside a JSON value. */
export function redactJson(value, unknown) {
  if (typeof value === 'string') return redact(value, unknown);
  if (Array.isArray(value)) return value.map((v) => redactJson(v, unknown));
  if (value && typeof value === 'object') return Object.fromEntries(Object.entries(value).map(([k, v]) => [k, redactJson(v, unknown)]));
  return value;
}
